CIScripts/Test/PesterLogger/RemoteLogCollector.ps1

```diff
@@ -41,23 +65,58 @@ function Get-LogContent {
         }
         return $Logs
     }
-    Invoke-CommandRemoteOrLocal -Func $ContentGetterBody -Session $LogSource.Session -Arguments $LogSource.Path
+    Invoke-CommandRemoteOrLocal -Func $ContentGetterBody -Session $Session -Arguments $Path
 }
 
-function Clear-LogContent {
-    param([System.Collections.Hashtable] $LogSource)
+function Clear-FileLogContent {
+    param([PSSessionT] $Session, [String] $Path)
     $LogCleanerBody = {
         Param([Parameter(Mandatory = $true)] [string] $What)
         $Files = Get-ChildItem -Path $What -ErrorAction SilentlyContinue
         foreach ($File in $Files) {
-            Remove-Item $File
+            try {
+                Remove-Item $File
+            }
+            catch {
+                Write-Warning "$File was not removed due to $_"
+            }
         }
     }
-    Invoke-CommandRemoteOrLocal -Func $LogCleanerBody -Session $LogSource.Session -Arguments $LogSource.Path
+    Invoke-CommandRemoteOrLocal -Func $LogCleanerBody -Session $Session -Arguments $Path
+}
+
+function Get-ContainerLogContent {
+    param([PSSessionT] $Session, [String] $ContainerName)
+    $Command = Invoke-NativeCommand -Session $Session -CaptureOutput -AllowNonZero {
+        docker logs $Using:ContainerName
+    }
+    @{
+        "$ContainerName container logs" = $Command.Output
+    }
+}
+
+function Get-LogContent {
+    param([LogSource] $LogSource)
+
+    if ($LogSource.Source['Path']) {
+        Get-FileLogContent -Session $LogSource.Session -Path $LogSource.Source.Path
+    } elseif ($LogSource.Source['Container']) {
+        Get-ContainerLogContent -Session $LogSource.Session -Container $LogSource.Source.Container
```

Comment:
Maybe something akin to a class would be better here? LogSource class with two subtypes: File and Container, with methods Clear and Get-Content.

1. Class inheritance would work given all those classes (LogSource, File and Container) are in the same file.
2. Otherwise, New-*LogSource functions could return structs with fields pointing to relevant functions (like in one of early prototypes of RemoteLogCollector)

I wouldn't do any of the above if it's too problematic though.

Reply:
I've tried to simulate tagged union with match in powershell and perhaps it may look weird. I'd argue that it's readable enough.

Solution 2. would have low readability (hard to track where's the code that actually runs), so I'd prefer not to do it.

Solution 1. seems nice, but do we want everybody to learn yet another feature of powershell just to do a simple switch? I'll try to implement it using classes but not sure if it's worth the time.

Reply:
> tagged union 
Very procedural, not OOP. However, without classes we probably are better off sticking to it. Notice how adding a new log source method requires changing `function Get-LogContent`. It should've been handled by polymorphysm. But classes in PS don't work too well so.. meh.

> learn yet another feature of powershell just to do a simple switch

but this is not a valid reason imo.

> Solution 2. would have low readability (hard to track where's the code that actually runs), so I'd prefer not to do it.

This is a tradeoff between maintainability and explicitness. Switch-case is more explicit but harder to maintain in the long run, because adding (extending) the system requires changes in unrelated (client) parts. 

I believe by removing the "closures" that were in the original (rejected) implementaiton of RemoteLogCollector, and now by using a switch-case, we're introducing a tech debt. Imagine this switch-case having 10 fields, for different log sources... A nightmare.

The only reason it's OK in this case is because powershell polymorphysm sucks.